File: client-android/serenada-call-ui/src/main/java/app/serenada/callui/CallRendererProvider.kt
package app.serenada.callui

import org.webrtc.SurfaceViewRenderer

interface CallRendererProvider {
    fun attachLocalRenderer(renderer: SurfaceViewRenderer)
    fun detachLocalRenderer(renderer: SurfaceViewRenderer)
    fun attachRemoteRenderer(renderer: SurfaceViewRenderer)
    fun detachRemoteRenderer(renderer: SurfaceViewRenderer)
    fun attachRemoteRenderer(renderer: SurfaceViewRenderer, cid: String)
    fun detachRemoteRenderer(renderer: SurfaceViewRenderer, cid: String)

    // Independent CONTENT (screen share) renderers. Default no-ops keep existing
    // providers source-compatible; the bundled flow supplies real wiring.
    fun attachRemoteContentRenderer(renderer: SurfaceViewRenderer, cid: String) {}
    fun detachRemoteContentRenderer(renderer: SurfaceViewRenderer, cid: String) {}
    fun attachLocalContentRenderer(renderer: SurfaceViewRenderer) {}
    fun detachLocalContentRenderer(renderer: SurfaceViewRenderer) {}
}

File: client-ios/SerenadaCallUI/Sources/CallRendererProvider.swift
import Foundation

@MainActor
protocol CallRendererProvider: AnyObject {
    func attachLocalRenderer(_ renderer: AnyObject)
    func detachLocalRenderer(_ renderer: AnyObject)
    func attachRemoteRenderer(_ renderer: AnyObject)
    func detachRemoteRenderer(_ renderer: AnyObject)
    func attachRemoteRenderer(_ renderer: AnyObject, forCid cid: String)
    func detachRemoteRenderer(_ renderer: AnyObject, forCid cid: String)

    // Independent CONTENT (screen share) renderers. The camera renderers above
    // stay on the camera track; these target the separate content track exposed
    // by the SDK in independent mode. They are no-ops when the SDK is flag-off
    // (no content track exists) so attaching them is always safe.
    func attachLocalContentRenderer(_ renderer: AnyObject)
    func detachLocalContentRenderer(_ renderer: AnyObject)
    func attachRemoteContentRenderer(_ renderer: AnyObject, forCid cid: String)
    func detachRemoteContentRenderer(_ renderer: AnyObject, forCid cid: String)
}
